### PR TITLE
Automatically fix unexportable objects so they can be exported normally

### PR DESCRIPTION
**Pull request by X123M3-256**
Certain objects created with Buggy's ridemaker have some nonzero bytes that should be zero. When the game loads the object, these bytes are zeroed, but the checksum is not updated. The invalid object is then written to the file. These objects then won't unpack correctly because the checksum doesn't match, requiring the user to install the problem object manually.

I have modified the loading code so that object checksums are ignored when "allow loading files with invalid checksums" is enabled. This allows to load parks containing unexportable objects without having to install the object manually.

I have also made it so that objects being written are automatically fixed so that they export correctly, even when loaded in the original game. This is done by appending some extra bytes to the file being written, so that the checksum matches whats in the file. This way it doesn't modify the checksum, so the game won't treat it as a different object.

### CODE REVIEW

File: src/object.c [hunk 1]
```diff
@@ -208,10 +209,56 @@ int write_object_file(SDL_RWops *rw, rct_object_entry* entry)
...
+		into the old buffer. If the chunk is only referenced in one place it would be safe to realloc 
+		it and update that reference, but I don't know the codebase well enough to know if that's the
+		case, so to be on the safe side I copy it*/
+		uint8* new_chunk=malloc(chunkHeader.length+11);
+		memcpy(new_chunk,chunk,chunkHeader.length);
+		//It should be safe to update these in-place because they are local
+		chunk=new_chunk;
+		chunkHeader.length+=11;
+
+		printf("New chunk size %d\n",chunkHeader.length);
```
Comment (janisozaur) [inline src/object.c]: `log_verbose`, same in other places

File: src/object.c [hunk 2]
```diff
@@ -208,10 +209,56 @@ int write_object_file(SDL_RWops *rw, rct_object_entry* entry)
...
+
+		printf("Fixing object with invalid checksum\n");
+		//Store the current length of the header - it's the offset at which we will write the extra bytes
+		int salt_offset=chunkHeader.length;
+		/*Allocate a new chunk 11 bytes longer.
+		I would just realloc the old one, but realloc can move the data, leaving dangling pointers 
+		into the old buffer. If the chunk is only referenced in one place it would be safe to realloc 
+		it and update that reference, but I don't know the codebase well enough to know if that's the
+		case, so to be on the safe side I copy it*/
+		uint8* new_chunk=malloc(chunkHeader.length+11);
```
Comment (janisozaur) [inline src/object.c]: This never gets `free`d

File: src/object.c [hunk 3]
```diff
@@ -208,10 +209,56 @@ int write_object_file(SDL_RWops *rw, rct_object_entry* entry)
 	chunkHeader.encoding = object_entry_group_encoding[type];
 	chunkHeader.length = installed_entry->chunk_size;
 
+
+	//Check if content of object file matches the stored checksum. If it does not, then fix it.
+	int calculated_checksum=object_calculate_checksum(entry,chunk,installed_entry->chunk_size);
+		if(entry->checksum!=calculated_checksum)
```
Comment (janisozaur) [inline src/object.c]: please indent properly

File: src/object.c (hunk 2 above)
Comment (X123M3-256) [inline src/object.c]: Oops... because I initially used realloc(), I forgot to add it. Should I submit a different pull request or is there a way to change this one?

Comment (janisozaur) [inline src/object.c]: If you push to the same branch as this PR, github will automatically update the PR.

You can amend your commit (and `git push -f`) or add another one, if you wish.

Comment (X123M3-256) [inline src/object.c]: OK, I'll do that now then

Comment (X123M3-256) [inline src/object.c]: OK, hopefully that should have fixed it. It now frees the modified chunk immediately after writing it.

File: src/object.c [hunk 4]
```diff
@@ -208,9 +209,58 @@ int write_object_file(SDL_RWops *rw, rct_object_entry* entry)
 	chunkHeader.encoding = object_entry_group_encoding[type];
 	chunkHeader.length = installed_entry->chunk_size;
 
-	size_dst += sawyercoding_write_chunk_buffer(dst_buffer + sizeof(rct_object_entry), chunk, chunkHeader);
-	SDL_RWwrite(rw, dst_buffer, 1, size_dst);
 
+	//Check if content of object file matches the stored checksum. If it does not, then fix it.
+	int calculated_checksum=object_calculate_checksum(entry,chunk,installed_entry->chunk_size);
+		if(entry->checksum!=calculated_checksum)
```
Comment (janisozaur) [inline src/object.c]: Please fix indentation of the changes.

File: src/object.c [hunk 5]
```diff
@@ -208,9 +209,58 @@ int write_object_file(SDL_RWops *rw, rct_object_entry* entry)
...
+		{
+		printf("Chunk size %d\n",chunkHeader.length);
```
Comment (janisozaur) [inline src/object.c]: You should be using `log_verbose` instead of `printf`

Comment (X123M3-256) [inline src/object.c]: Those printfs were all supposed to be removed. I'll do that and make another commit.

File: src/object.c (hunk 4 above)
Comment (X123M3-256) [inline src/object.c]: What's wrong with the indentation?

File: src/object.c (hunk 5 above)
Comment (X123M3-256) [inline src/object.c]: OK, so I've removed all the spurious printfs.

File: src/object.c (hunk 4 above)
Comment (janisozaur) [inline src/object.c]: Indentation suggests how the code flows, so a child block is more indented than the parent.

```
// parent scope
int x = 3;
if (x == 2) {
    // indented child scope
    x++;
} else {
    x--;
}
```

Comment (X123M3-256) [inline src/object.c]: OK, I'll change it to match that then.

**Comment by janisozaur** [comment]
Review :+1: from me.

Tested it with http://www.nedesigns.com/park/3482/busch-gardens-asia/ park, exported the park with fixed version, removed `RIDETESJ.DAT` from `ObjData/`, opened exported version, worked just as expected.

**Comment by IntelOrca** [comment]
I don't normally enforce code style with C code as it will all change eventually but, all this code is very different from the rest of the code base. Notably there is a lack of spaces around any of the symbols. Could you change it so that all operators (e.g. = or &&) are surround with spaces?

**Comment by X123M3-256** [comment]
OK, I have added spaces around the operators.

File: src/object.c [hunk 6]
```diff
@@ -237,9 +279,14 @@ int object_load_packed(SDL_RWops* rw)
 	}
 
 	if (object_calculate_checksum(&entry, chunk, chunkSize) != entry.checksum){
-		log_error("Checksum mismatch from packed object: %.8s", entry.name);
-		free(chunk);
-		return 0;
+	
+		if(gConfigGeneral.allow_loading_with_incorrect_checksum){
```
Comment (wlhlm) [inline src/object.c]: Space

File: src/object.c [hunk 7]
```diff
@@ -208,9 +209,50 @@ int write_object_file(SDL_RWops *rw, rct_object_entry* entry)
 	chunkHeader.encoding = object_entry_group_encoding[type];
 	chunkHeader.length = installed_entry->chunk_size;
 
-	size_dst += sawyercoding_write_chunk_buffer(dst_buffer + sizeof(rct_object_entry), chunk, chunkHeader);
-	SDL_RWwrite(rw, dst_buffer, 1, size_dst);
 
+	//Check if content of object file matches the stored checksum. If it does not, then fix it.
+	int calculated_checksum=object_calculate_checksum(entry,chunk,installed_entry->chunk_size);
+	if(entry->checksum != calculated_checksum){
```
Comment (wlhlm) [inline src/object.c]: Space

File: src/object.c [hunk 8]
```diff
@@ -208,9 +209,50 @@ int write_object_file(SDL_RWops *rw, rct_object_entry* entry)
...
+		chunkHeader.length += 11;
+
+		/*Next work out which bits need to be flipped to make the current checksum match the one in the file 
+		The bitwise rotation compensates for the rotation performed during the checksum calculation*/
+		int bits_to_flip = entry->checksum ^ ((calculated_checksum << 25) | (calculated_checksum >> 7));
+		/*Each set bit encountered during encoding flips one bit of the resulting checksum (so each bit of the checksum is an XOR
+		of bits from the file). Here, we take each bit that should be flipped in the checksum and set one of the bits in the data 
+		that maps to it. 11 bytes is the minimum needed to touch every bit of the checksum - with less than that, you wouldn't 
+		always be able to make the checksum come out to the desired target*/
+		new_chunk[salt_offset] = (bits_to_flip & 0x00000001) << 7;;
```
Comment (wlhlm) [inline src/object.c]: Double semicolon

File: src/object.c (hunk 7 above)
Comment (X123M3-256) [inline src/object.c]: OK, I fixed this now

File: src/object.c (hunk 8 above)
Comment (X123M3-256) [inline src/object.c]: Now fixed

File: src/object.c (hunk 6 above)
Comment (X123M3-256) [inline src/object.c]: Where should the spaces go here?

Comment (wlhlm) [inline src/object.c]: After `if` and before the curly bracket, i.e.
```c
if (gConfigGeneral.allow_loading_with_incorrect_checksum) {
```

Same thing in [line 215](https://github.com/OpenRCT2/OpenRCT2/pull/3021/files#diff-43278b647ec38f9faf284496e22a97d5R215)

Comment (X123M3-256) [inline src/object.c]: OK, should be fixed now

**Comment by janisozaur** [comment]
Many thanks for your work @X123M3-256

**Comment by X123M3-256** [comment]
Is there anything else that needs fixing?

**Comment by Gymnasiast** [comment]
Well, all issues labelled 'bug', 'original bug' and many labelled 'investigate'. ;-)
Thank you for this one!

**Comment by X123M3-256** [comment]
I was talking about this specific code - there were a lot of problems with it, and I only had today to fix them. I won't be able to make any more changes until next weekend, so if there's any more missing spaces or stray printfs, I can't change them for a while.

**Comment by birthdaybrian** [comment]
I'm still getting the Checksum mismatch from packed object message. I will host a server to see if anyone can join my server when I get the message in the command prompt. Two of the rides are Amazing Earl rides which are the swimming pool and the waterslide.

**Comment by janisozaur** [comment]
The fix to objects is only applied on saving a park w/ exported objects. Make sure you have that option checked, save and reload your park.

It should also be applied when sending a map over network (which does an implicit save) with the option enabled, but you will still get the message as you are loading unfixed park.

**Comment by birthdaybrian** [comment]
@janisozaur I have exporting checked.

**Comment by X123M3-256** [comment]
Can you send me the save file you have?

**Comment by birthdaybrian** [comment]
Many people except for two are having problems joining my server with too many custom objects.
Here is the save game: http://1drv.ms/21Gli4J This save file takes a whopping 26MB.

**Comment by X123M3-256** [comment]
It's a bit late, but I can confirm there is an issue. Resaving the game fixed RIDE9MKY but MetRTube and MetRsurf still don't export correctly. The objects are definitely included with the save because I can open the park if I select "allow loading with incorrect checksums".

I have been able to reproduce the issue with a fresh park containing only MetRsurf. It does not work. When I have some more time available I'll try to find out why. For now, you can select "allow loading with incorrect checksums" to open the park but bear in mind that this results in an invalid object being saved to disk (unless you already have the object), in which case you should be able to open it anyway.
